src/Interpreters/ArithmeticOperationsInAgrFuncOptimize.cpp

```diff
@@ -0,0 +1,286 @@
+                        ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+    auto * inter_node = f_n->arguments->children[0]->as<ASTFunction>();
+    if (inter_node && inter_node->name == mul)
+    {
+        int tp = 1;
+        std::pair<ASTs, ASTs> nodes = findAllConsts(f_n, mul);
+
+        if (nodes.first.empty())
+            return;
+
+        for (const auto & arg : nodes.first)
+        {
+            auto num = arg->as<ASTLiteral>()->value.get<Int128>();
+
+            /// if multiplication is negative, min function becomes max
+
+            if ((arg->as<ASTLiteral>()->value.getType() == Field::Types::Int64 ||
+                 arg->as<ASTLiteral>()->value.getType() == Field::Types::Int128) && static_cast<int64_t>(num) < 0)
+                tp *= -1;
+        }
+
+        std::pair<int, std::string> cur = {tp, max};
+        buildTree(f_n, min, mul, nodes, cur);
+    }
+    else if (inter_node && inter_node->name == plus)
+    {
+        std::pair<ASTs, ASTs> nodes = findAllConsts(f_n, plus);
+        std::pair<int, std::string> cur = {1, min};
+        buildTree(f_n, min, plus, nodes, cur);
+    }
+}
+
+void maxOptimize(ASTFunction * f_n)
+{
+    std::string max = "max";
+    std::string min = "min";
+    std::string mul = "multiply";
+    std::string plus = "plus";
+
+    const auto * function_args = f_n->arguments->as<ASTExpressionList>();
+
+    if (!function_args || function_args->children.size() != 1)
+        throw Exception("Wrong number of arguments for function 'max' (" + toString(function_args->children.size()) + " instead of 1)",
+                        ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+    auto * inter_node = f_n->arguments->children[0]->as<ASTFunction>();
+    if (inter_node && inter_node->name == mul)
+    {
+        int tp = 1;
+        std::pair<ASTs, ASTs> nodes = findAllConsts(f_n, mul);
+
+        if (nodes.first.empty())
+            return;
+
+        for (const auto &ar: nodes.first)
+        {
+            auto num = ar->as<ASTLiteral>()->value.get<Int128>();
+
+            /// if multiplication is negative, max function becomes min
+            if ((ar->as<ASTLiteral>()->value.getType() == Field::Types::Int64 ||
+                    ar->as<ASTLiteral>()->value.getType() == Field::Types::Int128) && static_cast<int64_t>(num) < 0)
+                tp *= -1;
+        }
+
+        std::pair<int, std::string> cur = {tp, min};
+        buildTree(f_n, max, mul, nodes, cur);
+    }
+    else if (inter_node && inter_node->name == plus)
+    {
+        std::pair<ASTs, ASTs> nodes = findAllConsts(f_n, plus);
+        std::pair<int, std::string> cur = {1, max};
+        buildTree(f_n, max, plus, nodes, cur);
+    }
+}
+
+/// optimize for min, max, sum is ready, ToDo: any, anyLast, groupBitAnd, groupBitOr, groupBitXor
+void ArithmeticOperationsInAgrFuncVisitor::visit(ASTPtr & current_ast)
```

Comment:
Too complex visit(child) logic. Rewrite with InDepthNodeVisitor. It realize stateless traversal througth tree. So there's no way to make mistake there.